### PR TITLE
Covered issue #646

### PR DESCRIPTION
**Pull request by Himanshunitrr**

### CODE REVIEW

File: core/main/src/Core/Container.ts [hunk 1]
```diff
@@ -37,6 +38,7 @@ export class Container {
     public started: boolean;
     public destroyed: boolean;
     public density: number;
+    public themes: Theme | ITheme[];
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts]: This property is not needed

File: core/main/src/Core/Container.ts [hunk 2]
```diff
@@ -393,4 +396,9 @@ export class Container {
 
         this.density = (canvas.width * canvas.height) / (densityOptions.factor * pxRatio * densityOptions.area);
     }
+
+    private loadTheme(): void {
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts]: This method needs a `name?: string` property, when passed searches the theme in the options with that name, when `undefined` (not passed) searches the `default` theme in the options

File: core/main/src/Options/Classes/Options.ts [hunk 3]
```diff
@@ -51,7 +53,7 @@ export class Options implements IOptions, IOptionLoader<IOptions> {
     public particles: Particles;
     public pauseOnBlur: boolean;
     public preset?: string | string[];
-
+    public themes: Theme | ITheme[];
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Options.ts]: This type should be `Theme[]`

File: core/main/src/Options/Classes/Options.ts [hunk 4]
```diff
@@ -62,6 +64,7 @@ export class Options implements IOptions, IOptionLoader<IOptions> {
         this.interactivity = new Interactivity();
         this.particles = new Particles();
         this.pauseOnBlur = true;
+        this.themes = new Theme();
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Options.ts]: this assignment should be only `[]`

File: core/main/src/Options/Classes/Options.ts [hunk 5]
```diff
@@ -102,12 +105,12 @@ export class Options implements IOptions, IOptionLoader<IOptions> {
         if (data.pauseOnBlur !== undefined) {
             this.pauseOnBlur = data.pauseOnBlur;
         }
-
         this.background.load(data.background);
         this.particles.load(data.particles);
         this.infection.load(data.infection);
         this.interactivity.load(data.interactivity);
         this.backgroundMask.load(data.backgroundMask);
+        this.themes.load(data.themes);
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Options.ts]: This needs a for over `data.themes` if not `undefined`

File: core/main/src/Options/Classes/Theme/Theme.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import type {ITheme} from "../../Interfaces/Theme/ITheme"
+import type { RecursivePartial } from "../../../Types";
+import {IOptionLoader} from "../../Interfaces/IOptionLoader"
+export class Theme implements ITheme, IOptionLoader<ITheme>{
+   
+    public name: string;
+    public default: boolean;
+
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Theme/Theme.ts]: Here's missing a property `public options?: RecursivePartial<IOptions>`

File: core/main/src/Options/Classes/Theme/Theme.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        if (data.name !== undefined) {
+            this.name = data.name;
+        }
+
+        if (data.default !== undefined) {
+            this.default = data.default;
+        }
+    }
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Theme/Theme.ts:31]: This is needed the load of the `options` property described above like the file `Trail.ts` does for `particles` property, located under Options Interactivity Modes

File: core/main/src/Options/Interfaces/IOptions.ts [hunk 8]
```diff
@@ -67,4 +69,7 @@ export interface IOptions {
      * @deprecated use the new detectRetina instead
      */
     retina_detect: boolean;
+    
+    // Optional theme
+    themes?: Theme | ITheme[];
```
Comment (matteobruni) [inline core/main/src/Options/Interfaces/IOptions.ts]: this type should be `ITheme[]`

**Review by matteobruni — COMMENTED** [review]
Review done

File: core/main/src/Options/Classes/Options.ts (hunk 5 above)
Comment (Himanshunitrr) [reply core/main/src/Options/Classes/Options.ts:113]: sth like
 if(data.themes !== undefined){
            for(const theme in data.themes){
                this.themes.push(theme.load(themes))
            }
        }

File: core/main/src/Options/Classes/Options.ts [hunk 9]
```diff
@@ -102,12 +105,16 @@ export class Options implements IOptions, IOptionLoader<IOptions> {
...
             this.pauseOnBlur = data.pauseOnBlur;
         }
-
         this.background.load(data.background);
         this.particles.load(data.particles);
         this.infection.load(data.infection);
         this.interactivity.load(data.interactivity);
         this.backgroundMask.load(data.backgroundMask);
+        if(data.themes !== undefined){
+            for(const theme in data.themes){
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Options.ts]: this should be `for (const theme of data.themes) {`

`in` enumerates property, `of` enumerates the array

File: core/main/src/Options/Classes/Options.ts [hunk 10]
```diff
@@ -102,12 +105,16 @@ export class Options implements IOptions, IOptionLoader<IOptions> {
...
+                this.themes.load(theme)
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Options.ts]: this should be replaced with this code:

```javascript
const optTheme = new Theme();

optTheme.load(theme);

this.themes.push(optTheme);
```

File: core/main/src/Options/Classes/Theme/Theme.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import {IOptions} from "../../Interfaces/IOptions"
+import {Utils} from "../../../Utils"
+
+export class Theme implements ITheme, IOptionLoader<ITheme>{
+   
+    public name: string;
+    public default: boolean;
+    public options?: RecursivePartial<IOptions>
+    constructor() {
+        this.name = "dark";
```
Comment (matteobruni) [inline core/main/src/Options/Classes/Theme/Theme.ts]: defaults to empty string

File: core/main/src/Options/Interfaces/IOptions.ts [hunk 12]
```diff
@@ -67,4 +69,7 @@ export interface IOptions {
      * @deprecated use the new detectRetina instead
      */
     retina_detect: boolean;
+    
+    // Optional theme
+    themes?: ITheme[];
```
Comment (matteobruni) [inline core/main/src/Options/Interfaces/IOptions.ts]: The `?` can be removed

File: core/main/src/Core/Container.ts [hunk 13]
```diff
@@ -393,4 +393,11 @@ export class Container {
 
         this.density = (canvas.width * canvas.height) / (densityOptions.factor * pxRatio * densityOptions.area);
     }
+
+    private loadTheme(name?: string): void {
+        if(name){
+        for(const themeOption in this.options.themes.)
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts]: `for (const themeOption of this.options.themes) {` is the right way, but a shorter way to find that would be nicer like `filter` or `find`

File: core/main/src/Core/Container.ts [hunk 14]
```diff
@@ -393,4 +392,13 @@ export class Container {
 
         this.density = (canvas.width * canvas.height) / (densityOptions.factor * pxRatio * densityOptions.area);
     }
+
+    private loadTheme(name?: string): void {
+        if(name){
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts:398]: This if needs an else searching for the default theme using find like for name

**Review by matteobruni — CHANGES_REQUESTED** [review]

File: core/main/src/Core/Container.ts [hunk 15]
```diff
@@ -393,4 +392,13 @@ export class Container {
 
         this.density = (canvas.width * canvas.height) / (densityOptions.factor * pxRatio * densityOptions.area);
     }
+
+    private loadTheme(name?: string): void {
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts]: This should be public, users can switch theme anytime they want

File: core/main/src/Core/Container.ts (hunk 14 above)
Comment (Himanshunitrr) [reply core/main/src/Core/Container.ts:398]: But default name string is empty.

Comment (matteobruni) [reply core/main/src/Core/Container.ts:398]: yes, but there's a default property on Theme class, you have to search for the one with true value

File: core/main/src/Core/Container.ts [hunk 16]
```diff
@@ -393,12 +394,15 @@ export class Container {
...
 
-    private loadTheme(name?: string): void {
+    public loadTheme(name?: string): void {
         if(name){
            let choosenTheme = this.options.themes.find(theme => theme.name === name)
             if(choosenTheme){
                 this.options.load(choosenTheme.options)
             }
+        }else{
+            let defaultTheme = new Theme()
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts:404]: this should be something like: this.options.themes.find(theme => theme.default)

File: core/main/src/Core/Container.ts [hunk 17]
```diff
@@ -16,6 +16,7 @@ import type { INoiseValue } from "./Interfaces/INoiseValue";
 import type { INoise } from "./Interfaces/INoise";
 import type { IRgb } from "./Interfaces/IRgb";
 import type { IAttract } from "./Interfaces/IAttract";
+import { Theme } from '../Options/Classes/Theme/Theme';
```
Comment (matteobruni) [inline core/main/src/Core/Container.ts:19]: This import is not necessary

**Review by matteobruni — APPROVED** [review]

**Pull request merged**